[William Ming] iP

cpwill01 opened the pull request:
# Duke

> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Duke keeps track of your tasks to free your mind of the clutter. It's
* text-based
* easy to learn
* super fast to use
* and it's **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders

Use Duke today in 4 simple steps:
1. Download ***duke.jar*** from the latest release [here](https://github.com/cpwill01/ip/releases)
2. Double click the downloaded file
3. Add your tasks
4. Let Duke manage your tasks for you 🎉 
---
### Some extra stuff
If you're a Java programmer, you can use it to practice java too. Here's the `main()` method:
```java
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
```

Commit:
Add Gradle support

cpwill01 committed:
Completed Level 1

cpwill01 committed:
Completed Level 2

cpwill01 committed:
Implement Task and Tasklist classes

cpwill01 committed:
Complete Level 3

cpwill01 committed:
Complete Level 4

cpwill01 committed:
Add test cases

Tests basic functionalities without error handling

cpwill01 committed:
Complete Level 5

cpwill01 committed:
Add test cases for error handling

cpwill01 committed:
Complete Level 6

cpwill01 committed:
Complete Level 7 - Saving & loading of tasks

cpwill01 committed:
Change deadline and event dates to LocalDate

cpwill01 committed:
Fix merge conflict from merging Level-8 to master

cpwill01 committed:
Extract Storage and parts of Ui from Duke.java

cpwill01 committed:
Extract Parser and Command classes from Duke.java

cpwill01 committed:
Reorganise classes into packages

cpwill01 committed:
Merge branch 'add-gradle-support'

cpwill01 committed:
Add gradle support

cpwill01 committed:
Add JUnit tests for Parser and Deadline

cpwill01 committed:
Change mainClassName for gradle Shadow plugin

cpwill01 committed:
Add JavaDocs for public classes and methods.

cpwill01 committed:
Import checkstyle config files

cpwill01 committed:
Edit code to comply with gradle checkstyle

cpwill01 committed:
Add find command

cpwill01 committed:
Fix merge conflict from merging A-CodingStandard

cpwill01 committed:
Fix merge conflict from merging Level-9

cpwill01 committed:
Fix punctuation mistakes in comments

hsiaojietng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,48 @@
+import duke.task.TaskList;
+
+/**
+ * Encapsulates the Duke task manager bot.
```
Easy to follow header comment!

hsiaojietng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,48 @@
+    /**
+     * Starts the application by initialising and running a Duke bot.
+     *
+     * @param args command-line arguments, if any
```
Precise and simple comments

hsiaojietng commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,105 @@
+ * of the inputs supplied by the user.
+ */
+public class Parser {
+
```
You can create enum for these static variables as such if you want to: https://www.programiz.com/java-programming/enum-string

hsiaojietng commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,104 @@
+    /**
+     * Initialises a Storage object.
+     *
+     * @param dirPath path to the directory where the txt file is stored.
```
Provides the meaning for dirPath, nice!

hsiaojietng commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,158 @@
+     */
+    public void showGoodbye() {
+        System.out.println("Bye! Hope to see you again!");
+    }
```
Very easy to follow the behaviours of these functions!

hsiaojietng commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,72 @@
+ */
+public abstract class Task {
+    private String description;
+    private boolean isDone;
```
Good naming convention!

hsiaojietng reviewed (commented):
LGTM. There are nits that you can check out but not necessary to make changes if not needed.

cpwill01 committed:
Add a basic GUI

cpwill01 committed:
Fix typos and coding standard violations.

joeietao commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,68 @@
+    private TaskList taskList;
+    private Storage storage;
+    private Ui ui;
+    private boolean isExit;
```
Overall I think your code quality is really good! You could consider simplifying isExit tho, I think its a little complicated for it to just need to exit the app. Could consider just exiting directly when user types bye/exit instead of having to change the state of isExit in so many classes

joeietao commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("[D]")
```
Maybe you could do String + String and so on instead of append to make it easier to read

joeietao commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,141 @@
+ * Encapsulates the text user interface of Duke, which deals
+ * with providing the messages sent from Duke to the user.
+ */
+public class Ui {
```
I really like the names of the methods in this class, they are all descriptive and intuitive!

joeietao reviewed (commented):
LGTM!! I think your code is easy to understand and follow, you followed the naming conventions really well, I could learn more from you HAHAHA 

If anything I think you could be more straightforward in some implementations, for eg the toString method doesnt really need a Stringbuilder, you can just concatenate the strings HAHA The isExit function was kinda confusing for me too, I didn't really understand why the program doesn't just exit immediately when the user inputs bye. 

Other than that, I think your code quality was really good, I learnt a lot!:)

SimJM commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,68 @@
+     * @param dirPath path to the directory where the txt file is stored.
+     * @param fileName name of the txt file.
+     */
+    public Duke(String dirPath, String fileName) {
```
The full spelling of dirPath could be a better naming for this parameter.

SimJM commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,105 @@
+        String commandString = inputArr[0];
+        String details = inputArr.length > 1 ? inputArr[1] : "";
+        try {
+            switch (commandString) {
```
Clear naming and good usage of switch

SimJM commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Encapsulates a command to be executed in Duke.
+ */
+public abstract class Command {
```
Documentation is very clear.

SimJM commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+     * @return the string representation of this Deadline task.
+     */
+    @Override
+    public String toString() {
```
Another way to do this is to use String.format() 
refer to : https://www.javatpoint.com/java-string-format

SimJM commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,68 @@
+/**
+ * Encapsulates the Duke task manager bot.
+ *
+ * @author William Ming
```
Overall coding standard and naming is well done, no major faults found. Keep up the good work!

cpwill01 committed:
Add application name and icon.

When running the .jar file, users are unable to see the icon or name of
the application.

Let's add an app icon and an app name that users can see in their dock
when using the application.

cpwill01 committed:
Add assertions to various classes

cpwill01 committed:
Extract methods and replace magic literals

The Storage class has long loadData and saveData methods which also use
magic literals.

Extracting the logic in these methods into separate methods improves the
code quality by shortening the methods and adhering better to the SLAP
principle.

Replacing magic literals with well-named constants also improves
readability.

Let's
* extract methods to handle the logic of loading and saving data
* replace magic literals with symbolic constants

cpwill01 committed:
Extract methods in Parser class

The parse method in Parser is long, has considerable nesting and mixees
different levels of abstraction.

To improve code quality, let's extract the logic for creating Todos,
Deadlines and Events from a given user input into 1 method each.

cpwill01 committed:
Merge pull request #1 from cpwill01/branch-A-Assertions

Add assertions to various classes

cpwill01 committed:
Fix merge commit from merging master

cpwill01 committed:
Merge pull request #2 from cpwill01/branch-A-CodeQuality

Branch-A-CodeQuality

cpwill01 committed:
Implement remind command

cpwill01 committed:
Merge pull request #4 from cpwill01/B-Reminders

Implement remind command

cpwill01 committed:
Set up workflow for Continuous Integration

cpwill01 committed:
Fix coding standard violations

cpwill01 committed:
Fix bug in parsing user inputs for creating scheduled tasks

cpwill01 committed:
Improve the GUI

cpwill01 committed:
Improve behaviour of Find command

The Find command is case sensitive, which may be inconvenient for
users.

When the Find result is empty, Duke also shows the message
"Task list is empty". This may confuse users into thinking their entire
task list is empty when in reality only the find result is empty.

Let's
* make the Find command case insensitive
* change Duke's empty Find result message to be less confusing

cpwill01 committed:
Make commands case-insensitive

cpwill01 committed:
Change Duke's welcome message

Duke's welcome message immediately shows the full task list. This may
be unfocused and clutters the screen on startup, especially if the user
has a lot of tasks.

Let's make Duke's welcome message more focused by only showing a
reminder for tasks that are scheduled soon.

cpwill01 committed:
Update README and UserGuide

cpwill01 committed:
Update README.md

cpwill01 committed:
Set theme jekyll-theme-cayman

cpwill01 committed:
Add acknowledgement for user icon

cpwill01 committed:
Update UserGuide.md

cpwill01 committed:
Change markdown processor and add emoji support

cpwill01 committed:
Change welcome message

cpwill01 committed:
Change font due to issues on MacOS
